source/module_elecstate/potentials/H_TDDFT_pw.cpp

```diff
@@ -11,147 +14,282 @@ int H_TDDFT_pw::istep = -1;
-            // vext[ir] = vextold[ir]*2.74*cos(0.856*timenow)*sin(0.0214*timenow)*sin(0.0214*timenow)*0.01944;
-            vext[ir] = vextold[ir] * 2.74 * cos(w_h2 * timenow) * sin(w_h3 * timenow) * sin(w_h3 * timenow);
-        }
+        case 1:
+            vext_space[ir] = cal_v_space_length_potential(x);
+            break;
+
+        case 2:
+            vext_space[ir] = cal_v_space_length_potential(y);
+            break;
 
-        vl_pseudo[ir] += vext[ir];
+        case 3:
+            vext_space[ir] = cal_v_space_length_potential(z);
+            break;
 
-        // std::cout << "x: " << k <<"	" << "y: " << j <<"	"<< "z: "<< i <<"	"<< "ir: " << ir << std::endl;
-        // std::cout << "vext: " << vext[ir] << std::endl;
-        // std::cout << "vrs: " << vrs(is,ir) <<std::endl;
+        default:
+            std::cout << "direction of electric field is wrong" << endl;
+            break;
+        }
     }
-    std::cout << "For time step "<<H_TDDFT_pw::istep<<" :"<<" vext exists" << std::endl;
 
-    delete[] vextold;
-    delete[] vext;
+    ModuleBase::timer::tick("H_TDDFT_pw", "cal_v_space_length");
+    return;
+}
 
-    ModuleBase::timer::tick("H_TDDFT_pw", "cal_fixed_v");
+double H_TDDFT_pw::cal_v_space_length_potential(double i)
+{
+    double vext_space;
+    if (i < this->rho_basis_->nx * lcut1)
+    {
+        vext_space = ((i / this->rho_basis_->nx - lcut1) / (lcut1 + 1.0 - lcut2) - lcut1) * this->ucell_->lat0;
+    }
+    else if (i >= this->rho_basis_->nx * lcut1 && i < this->rho_basis_->nx * lcut2)
+    {
+        vext_space = -i / this->rho_basis_->nx * this->ucell_->lat0;
+    }
+    else if (i >= this->rho_basis_->nx * lcut2)
+    {
+        vext_space = ((i / this->rho_basis_->nx - lcut2) / (lcut1 + 1.0 - lcut2) - lcut2) * this->ucell_->lat0;
+    }
+    return vext_space;
+}
+
+void H_TDDFT_pw::cal_v_space_velocity(double* vext_space)
+{
     return;
-} // end subroutine set_vrs_tddft
+}
+
+double H_TDDFT_pw::cal_v_time()
+{
+    double vext_time = 0.0;
+
+    switch (ttype)
+    {
+    case 0:
+        vext_time = cal_v_time_Gauss();
+        cout << "gauss vtime=" << vext_time << endl;
+        break;
+
+    case 1:
+        vext_time = cal_v_time_trapezoid();
+        cout << "trape vtime=" << vext_time << endl;
+        break;
+
+    case 2:
+        vext_time = cal_v_time_trigonometric();
+        cout << "trigo vtime=" << vext_time << endl;
+        break;
+
+    case 3:
+        vext_time = cal_v_time_heaviside();
+        cout << "heavi vtime=" << vext_time << endl;
+        break;
+
+    case 4:
+        vext_time = cal_v_time_HHG();
+        cout << "hhg vtime=" << vext_time << endl;
+        break;
+
+    default:
+        std::cout << "time_domain_type of electric field is wrong" << endl;
+        break;
+    }
+    return vext_time;
+}
+
+double H_TDDFT_pw::cal_v_time_Gauss()
+{
+    double vext_time = 0.0;
+
+    double gauss_t = (istep - gauss_t0) * INPUT.mdp.md_dt;
```

Comment:
Don't use parameter of MD in TDDFT function, please use a internal variable to replace it.